lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp

```diff
@@ -193,8 +223,317 @@ struct LowerClassesPass : public LowerClassesBase<LowerClassesPass> {
   DenseMap<FModuleLike, bool> shouldCreateClassMemo;
 };
 
+struct PathTracker {
+  // An entry point for parallely tracking paths in the circuit and apply
+  // changes to `pathInfoTable`.
+  static LogicalResult
+  run(CircuitOp circuit, InstanceGraph &instanceGraph,
+      hw::InnerSymbolNamespaceCollection &namespaces, HierPathCache &cache,
+      PathInfoTable &pathInfoTable, const SymbolTable &symbolTable,
+      const DenseMap<DistinctAttr, FModuleOp> &owningModules);
+
+private:
+  PathTracker(FModuleLike module, hw::InnerSymbolNamespace &moduleNamespace,
+              InstanceGraph &instanceGraph, const SymbolTable &symbolTable,
+              const DenseMap<DistinctAttr, FModuleOp> &owningModules)
+      : module(module), moduleNamespace(moduleNamespace),
+        instanceGraph(instanceGraph), symbolTable(symbolTable),
+        owningModules(owningModules) {}
+
+  struct PathInfoTableEntry {
+    Operation *op;
+    DistinctAttr id;
+    StringAttr altBasePathModule;
+    // This is null if the path has no owning module.
+    ArrayAttr pathAttr;
+  };
+
+  LogicalResult runOnModule();
+  LogicalResult processPathTrackers(AnnoTarget target);
+  LogicalResult updatePathInfoTable(PathInfoTable &pathInfoTable,
+                                    HierPathCache &cache);
+
+  // Determine it is necessary to use an alternative base path for `moduleName`
+  // and `owningModule`.
+  FailureOr<bool> getOrComputeNeedsAltBasePath(Location loc,
+                                               StringAttr moduleName,
+                                               FModuleOp owningModule);
+  FModuleLike module;
+
+  // Local data structures.
+  hw::InnerSymbolNamespace &moduleNamespace;
+  DenseMap<std::pair<StringAttr, FModuleOp>, bool> needsAltBasePathCache;
+  SmallVector<Attribute> portAnnotations;
+
+  // Thread-unsafe global data structure. Don't mutate.
+  InstanceGraph &instanceGraph;
+  const SymbolTable &symbolTable;
+  const DenseMap<DistinctAttr, FModuleOp> &owningModules;
+
+  // Result.
+  SmallVector<PathInfoTableEntry> entries;
+  SetVector<StringAttr> altBasePathRoots;
+};
+
 } // namespace
 
+LogicalResult
+PathTracker::run(CircuitOp circuit, InstanceGraph &instanceGraph,
+                 hw::InnerSymbolNamespaceCollection &namespaces,
+                 HierPathCache &cache, PathInfoTable &pathInfoTable,
+                 const SymbolTable &symbolTable,
+                 const DenseMap<DistinctAttr, FModuleOp> &owningModules) {
+  SmallVector<PathTracker> trackers;
+  // Prepare workers.
+  for (auto module : circuit.getOps<FModuleLike>())
+    trackers.push_back(PathTracker(module, namespaces[module], instanceGraph,
+                                   symbolTable, owningModules));
```

Comment:
Can you use `emplace_back` here?

Might be slightly faster to iterate the instance graph here, to avoid scanning the hundreds of HierPathOps that are in the root of the circuit, but that is not an "optimization" that we have worried about before, and I'm not totally sure would help.

Reply:
Changed to used emplace_back and instance graph.